File: macos/Sources/FastCopyMac/APIClient.swift
import Foundation

enum APIClientError: LocalizedError {
    case invalidServerURL
    case invalidResponse
    case server(status: Int, code: String, message: String)

    var isUnauthorized: Bool {
        if case .server(let status, _, _) = self {
            return status == 401
        }
        return false
    }

    var errorDescription: String? {
        switch self {
        case .invalidServerURL:
            return "服务端地址无效"
        case .invalidResponse:
            return "服务端返回了无法识别的数据"
        case .server(_, _, let message):
            return message
        }
    }
}

struct APIClient {
    let baseURL: String
    private let session: URLSession
    private static let userAgent = "FastCopyMac/\(Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "0.2.1")"

    init(baseURL: String, session: URLSession = .shared) {
        self.baseURL = baseURL.trimmingCharacters(in: .whitespacesAndNewlines)
            .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        self.session = session
    }

    func authenticate(request: AuthRequest) async throws -> AuthResponse {
        try await send(
            AuthResponse.self,
            method: "POST",
            path: "/v1/auth/session",
            body: encode(request)
        )
    }

    func refresh(_ refreshToken: String) async throws -> RefreshResponse {
        struct Request: Encodable { let refreshToken: String }
        return try await send(
            RefreshResponse.self,
            method: "POST",
            path: "/v1/auth/refresh",
            body: encode(Request(refreshToken: refreshToken))
        )
    }

    func logout(token: String) async throws {
        try await sendWithoutResponse(method: "POST", path: "/v1/auth/logout", token: token)
    }

    func upload(_ upload: ClipUpload, token: String) async throws -> ClipCreateResponse {
        try await send(
            ClipCreateResponse.self,
            method: "POST",
            path: "/v1/clips",
            body: encode(upload),
            token: token
        )
    }

    func clips(afterSeq: Int64, token: String) async throws -> ClipsResponse {
        try await send(
            ClipsResponse.self,
            method: "GET",
            path: "/v1/clips",
            queryItems: [
                URLQueryItem(name: "after_seq", value: String(afterSeq)),
                URLQueryItem(name: "limit", value: "200")
            ],
            token: token
        )
    }

    func acknowledge(seq: Int64, token: String) async throws {
        struct Request: Encodable { let seq: Int64 }
        try await sendWithoutResponse(
            method: "POST",
            path: "/v1/sync/ack",
            body: encode(Request(seq: seq)),
            token: token
        )
    }

    func devices(token: String) async throws -> DevicesResponse {
        try await send(DevicesResponse.self, method: "GET", path: "/v1/devices", token: token)
    }

    func revokeDevice(id: String, token: String) async throws {
        try await sendWithoutResponse(
            method: "POST",
            path: "/v1/devices/\(id)/revoke",
            token: token
        )
    }

    func webSocketRequest(token: String) throws -> URLRequest {
        var components = try components(path: "/v1/events/ws", queryItems: [])
        switch components.scheme?.lowercased() {
        case "https": components.scheme = "wss"
        case "http": components.scheme = "ws"
        default: throw APIClientError.invalidServerURL
        }
        guard let url = components.url else { throw APIClientError.invalidServerURL }
        var request = URLRequest(url: url)
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue(Self.userAgent, forHTTPHeaderField: "User-Agent")
        return request
    }

    private func send<Response: Decodable>(
        _ type: Response.Type,
        method: String,
        path: String,
        queryItems: [URLQueryItem] = [],
        body: Data? = nil,
        token: String? = nil
    ) async throws -> Response {
        let data = try await sendData(
            method: method,
            path: path,
            queryItems: queryItems,
            body: body,
            token: token
        )
        do {
            return try Self.decoder.decode(type, from: data)
        } catch {
            throw APIClientError.invalidResponse
        }
    }

    private func sendWithoutResponse(
        method: String,
        path: String,
        queryItems: [URLQueryItem] = [],
        body: Data? = nil,
        token: String? = nil
    ) async throws {
        _ = try await sendData(
            method: method,
            path: path,
            queryItems: queryItems,
            body: body,
            token: token
        )
    }

    private func sendData(
        method: String,
        path: String,
        queryItems: [URLQueryItem],
        body: Data?,
        token: String?
    ) async throws -> Data {
        let components = try components(path: path, queryItems: queryItems)
        guard let url = components.url else { throw APIClientError.invalidServerURL }
        var request = URLRequest(url: url, timeoutInterval: 20)
        request.httpMethod = method
        request.httpBody = body
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue(Self.userAgent, forHTTPHeaderField: "User-Agent")
        if body != nil {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        if let token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw APIClientError.invalidResponse
        }
        guard (200..<300).contains(http.statusCode) else {
            if let envelope = try? Self.decoder.decode(APIErrorEnvelope.self, from: data) {
                throw APIClientError.server(
                    status: http.statusCode,
                    code: envelope.error.code,
                    message: envelope.error.message
                )
            }
            throw APIClientError.server(
                status: http.statusCode,
                code: "HTTP_\(http.statusCode)",
                message: "服务端请求失败（HTTP \(http.statusCode)）"
            )
        }
        return data
    }

    private func components(path: String, queryItems: [URLQueryItem]) throws -> URLComponents {
        guard var components = URLComponents(string: baseURL),
              let scheme = components.scheme?.lowercased(),
              ["http", "https"].contains(scheme),
              components.host != nil else {
            throw APIClientError.invalidServerURL
        }
        let prefix = components.path == "/" ? "" : components.path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        components.path = (prefix.isEmpty ? "" : "/\(prefix)") + path
        components.queryItems = queryItems.isEmpty ? nil : queryItems
        components.fragment = nil
        return components
    }

    private func encode<Value: Encodable>(_ value: Value) throws -> Data {
        try Self.encoder.encode(value)
    }

    private static let encoder: JSONEncoder = {
        let encoder = JSONEncoder()
        encoder.keyEncodingStrategy = .convertToSnakeCase
        return encoder
    }()

    private static let decoder: JSONDecoder = {
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        return decoder
    }()
}

File: windows/src/FastCopy.Core/ApiClient.cs
using System.Net;
using System.Net.Http.Json;
using System.Net.WebSockets;
using System.Text.Json;

namespace FastCopy.Core;

public sealed class FastCopyApiException : Exception
{
    public FastCopyApiException(HttpStatusCode statusCode, string code, string message)
        : base(message)
    {
        StatusCode = statusCode;
        Code = code;
    }

    public HttpStatusCode StatusCode { get; }
    public string Code { get; }
    public bool IsUnauthorized => StatusCode == HttpStatusCode.Unauthorized;
}

public sealed class FastCopyApiClient
{
    private const string UserAgent = "FastCopyWindows/0.1.1";
    private static readonly HttpClient Http = CreateHttpClient();
    private readonly Uri _baseUri;

    public FastCopyApiClient(string baseUrl)
    {
        var normalized = baseUrl.Trim().TrimEnd('/') + "/";
        if (!Uri.TryCreate(normalized, UriKind.Absolute, out var parsed)
            || (parsed.Scheme != Uri.UriSchemeHttps
                && !(parsed.Scheme == Uri.UriSchemeHttp && parsed.IsLoopback)))
        {
            throw new ArgumentException("服务端地址无效，非本机地址必须使用 HTTPS。", nameof(baseUrl));
        }
        _baseUri = parsed;
    }

    public Task<AuthResponse> AuthenticateAsync(AuthRequest request, CancellationToken cancellationToken) =>
        SendAsync<AuthResponse>(HttpMethod.Post, "v1/auth/session", null, request, cancellationToken);

    public Task<RefreshResponse> RefreshAsync(string refreshToken, CancellationToken cancellationToken) =>
        SendAsync<RefreshResponse>(
            HttpMethod.Post,
            "v1/auth/refresh",
            null,
            new RefreshRequest(refreshToken),
            cancellationToken);

    public Task LogoutAsync(string token, CancellationToken cancellationToken) =>
        SendWithoutResponseAsync(HttpMethod.Post, "v1/auth/logout", token, null, cancellationToken);

    public Task<ClipCreateResponse> UploadAsync(
        ClipUpload upload,
        string token,
        CancellationToken cancellationToken) =>
        SendAsync<ClipCreateResponse>(HttpMethod.Post, "v1/clips", token, upload, cancellationToken);

    public Task<ClipsResponse> ClipsAsync(
        long afterSeq,
        string token,
        CancellationToken cancellationToken) =>
        SendAsync<ClipsResponse>(
            HttpMethod.Get,
            $"v1/clips?after_seq={afterSeq}&limit=200",
            token,
            null,
            cancellationToken);

    public Task AcknowledgeAsync(long seq, string token, CancellationToken cancellationToken) =>
        SendWithoutResponseAsync(
            HttpMethod.Post,
            "v1/sync/ack",
            token,
            new AcknowledgeRequest(seq),
            cancellationToken);

    public Task<DevicesResponse> DevicesAsync(string token, CancellationToken cancellationToken) =>
        SendAsync<DevicesResponse>(HttpMethod.Get, "v1/devices", token, null, cancellationToken);

    public Task RevokeDeviceAsync(string deviceId, string token, CancellationToken cancellationToken) =>
        SendWithoutResponseAsync(
            HttpMethod.Post,
            $"v1/devices/{Uri.EscapeDataString(deviceId)}/revoke",
            token,
            null,
            cancellationToken);

    public async Task<ClientWebSocket> ConnectWebSocketAsync(
        string token,
        CancellationToken cancellationToken)
    {
        var socket = new ClientWebSocket();
        socket.Options.SetRequestHeader("Authorization", $"Bearer {token}");
        socket.Options.SetRequestHeader("User-Agent", UserAgent);
        socket.Options.KeepAliveInterval = TimeSpan.Zero;
        try
        {
            await socket.ConnectAsync(WebSocketUri(), cancellationToken).ConfigureAwait(false);
            return socket;
        }
        catch
        {
            socket.Dispose();
            throw;
        }
    }

    private async Task<T> SendAsync<T>(
        HttpMethod method,
        string path,
        string? token,
        object? body,
        CancellationToken cancellationToken)
    {
        using var response = await SendAsync(method, path, token, body, cancellationToken)
            .ConfigureAwait(false);
        await EnsureSuccessAsync(response, cancellationToken).ConfigureAwait(false);
        return await response.Content.ReadFromJsonAsync<T>(FastCopyJson.Options, cancellationToken)
            .ConfigureAwait(false)
            ?? throw new InvalidDataException("服务端返回了空响应。");
    }

    private async Task SendWithoutResponseAsync(
        HttpMethod method,
        string path,
        string? token,
        object? body,
        CancellationToken cancellationToken)
    {
        using var response = await SendAsync(method, path, token, body, cancellationToken)
            .ConfigureAwait(false);
        await EnsureSuccessAsync(response, cancellationToken).ConfigureAwait(false);
    }

    private async Task<HttpResponseMessage> SendAsync(
        HttpMethod method,
        string path,
        string? token,
        object? body,
        CancellationToken cancellationToken)
    {
        using var request = new HttpRequestMessage(method, new Uri(_baseUri, path));
        request.Headers.Accept.ParseAdd("application/json");
        request.Headers.UserAgent.ParseAdd(UserAgent);
        if (!string.IsNullOrEmpty(token))
        {
            request.Headers.Authorization = new("Bearer", token);
        }
        if (body is not null)
        {
            request.Content = JsonContent.Create(body, options: FastCopyJson.Options);
        }
        return await Http.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken)
            .ConfigureAwait(false);
    }

    private static async Task EnsureSuccessAsync(
        HttpResponseMessage response,
        CancellationToken cancellationToken)
    {
        if (response.IsSuccessStatusCode)
        {
            return;
        }
        ApiErrorEnvelope? envelope = null;
        try
        {
            envelope = await response.Content.ReadFromJsonAsync<ApiErrorEnvelope>(
                FastCopyJson.Options,
                cancellationToken).ConfigureAwait(false);
        }
        catch (JsonException)
        {
        }
        throw new FastCopyApiException(
            response.StatusCode,
            envelope?.Error.Code ?? $"HTTP_{(int)response.StatusCode}",
            envelope?.Error.Message ?? $"服务端请求失败（HTTP {(int)response.StatusCode}）。");
    }

    private Uri WebSocketUri()
    {
        var builder = new UriBuilder(new Uri(_baseUri, "v1/events/ws"))
        {
            Scheme = _baseUri.Scheme == Uri.UriSchemeHttps ? "wss" : "ws"
        };
        return builder.Uri;
    }

    private static HttpClient CreateHttpClient()
    {
        var handler = new SocketsHttpHandler
        {
            AutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate,
            ConnectTimeout = TimeSpan.FromSeconds(10),
            PooledConnectionIdleTimeout = TimeSpan.FromMinutes(2),
            PooledConnectionLifetime = TimeSpan.FromMinutes(10),
            MaxConnectionsPerServer = 4
        };
        return new HttpClient(handler) { Timeout = TimeSpan.FromSeconds(30) };
    }
}
